app/controllers/actions/AuthAction.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import config.EnrolmentKeys
+import models.auth.UserRequest
+import play.api.mvc.Results._
+import play.api.mvc._
+import uk.gov.hmrc.auth.core.AffinityGroup.Organisation
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import utils.Logging
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+trait AuthAction {
+  def apply(ern: String): ActionBuilder[UserRequest, AnyContent] with ActionFunction[Request, UserRequest]
+}
+
+@Singleton
+class AuthActionImpl @Inject()(override val authConnector: AuthConnector,
+                               val bodyParser: BodyParsers.Default
+                              )(implicit val ec: ExecutionContext) extends AuthAction with AuthorisedFunctions with Logging {
+
+  override def apply(ern: String): ActionBuilder[UserRequest, AnyContent] with ActionFunction[Request, UserRequest] =
+    new ActionBuilder[UserRequest, AnyContent] with ActionFunction[Request, UserRequest] {
+
+      override val parser = bodyParser
+      override val executionContext = ec
+
+      override def invokeBlock[A](request: Request[A], block: UserRequest[A] => Future[Result]): Future[Result] = {
+
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+
+        implicit val req = request
+
+        authorised().retrieve(Retrievals.affinityGroup and Retrievals.allEnrolments and Retrievals.internalId and Retrievals.credentials) {
+
+          case Some(Organisation) ~ enrolments ~ Some(internalId) ~ Some(credentials) =>
+            logger.debug("[invokeBlock] Checking for EMCS Enrolment")
+            checkOrganisationEMCSEnrolment(ern, enrolments, internalId, credentials.providerId)(block)
+
+          case Some(Organisation) ~ _ ~ None ~ _ =>
```

Comment:
I know this is quite a big statement already, but wondering if we should try to catch all the problems when logging, rather than one at a time?